yarn/src/main/scala/org/apache/spark/scheduler/cluster/YarnSchedulerBackend.scala

```diff
@@ -51,6 +51,9 @@ private[spark] abstract class YarnSchedulerBackend(
 
   private implicit val askTimeout = RpcUtils.askRpcTimeout(sc.conf)
 
+  // Flag to specify whether AM is registered or not.
+  @volatile private var isAmRegistered = false
```

Comment:
The name of this variable is misleading; it sounds like it's tracking whether there is an AM registered, but it's tracking whether an AM has ever been registered, even if currently there's no AM.

So the name should be something like "shouldResetOnAmRegistration" or something less verbose... which raises the question, could you just reset also on the first registration? What ill side-effects could that cause?